---
title: "DESEQ2 R Tutorial"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### RNA-Sequence Analysis Workflow
##### 1. Quality assess and clean raw sequencing data   
##### 2. Align reads to a reference  
##### 3. Count the number of reads assigned to each contig/gene    
##### 4. Extract counts and store in a matrix   
##### 5. Create column metadata table   
##### 6. Analyze count data using DESEQ2  
<br />
<br />
      
##### Install packages and load libraries

```{r message=FALSE}
#install.packages("htmltools")
#library(htmltools)
#source("https://bioconductor.org/biocLite.R")
#biocLite("DESeq2")

library( "DESeq2" )
library(ggplot2)


```
   
<br />
<br />

##### Download data   
```{r}
countsName <- "http://bioconnector.org/workshops/data/airway_scaledcounts.csv"
download.file(countsName, destfile = "airway_scaledcounts.csv", method = "auto")

countData <- read.csv('airway_scaledcounts.csv', header = TRUE, sep = ",")
head(countData)

metaDataName <- "http://bioconnector.org/workshops/data/airway_metadata.csv"
download.file(metaDataName, destfile = "airway_metadata.csv", method = "auto")

metaData <- read.csv('airway_metadata.csv', header = TRUE, sep = ",")
metaData

```

<br />
<br />

##### Construct DESEQDataSet Object   
```{r}
dds <- DESeqDataSetFromMatrix(countData=countData, 
                              colData=metaData, 
                              design=~dex, tidy = TRUE)
#Design specifies how the counts from each gene depend on our variables in the metadata
#For this dataset the factor we care about is our treatment status (dex)
#tidy=TRUE argument, which tells DESeq2 to output the results table with rownames as a first #column called 'row.

#let's see what this object looks like
dds

```


##### Now we're ready to run DESEQ function
```{r}
dds <- DESeq(dds)

```


<br />
<br />

##### What did we just do?
```{r}
?DESeq

#estimateSizeFactors
#This calculates the relative library depth of each sample 

#estimateDispersions
#estimates the dispersion of counts for each gene 

#nbinomWaldTest
#calculates the significance of coefficients in a Negative Binomial GLM using the size and dispersion outputs

```

<br />
<br />

##### Take a look at the results table

```{r}
res <- results(dds)
head(results(dds, tidy=TRUE)) #let's look at the results table

```


<br />
<br />

##### Summary of differential gene expression

```{r}
summary(res) #summary of results

```



<br />
<br />

##### Sort summary list by p-value

```{r}
res <- res[order(res$padj),]
head(res)

```
<br />
<br />

##### plotCounts

```{r}
#we can use plotCounts fxn to compare the normalized counts
#between treated and control groups for our top 6 genes
par(mfrow=c(2,3))

plotCounts(dds, gene="ENSG00000152583", intgroup="dex")
plotCounts(dds, gene="ENSG00000179094", intgroup="dex")
plotCounts(dds, gene="ENSG00000116584", intgroup="dex")
plotCounts(dds, gene="ENSG00000189221", intgroup="dex")
plotCounts(dds, gene="ENSG00000120129", intgroup="dex")
plotCounts(dds, gene="ENSG00000148175", intgroup="dex")

#Next steps in exploring these data...BLAST to database to find associated gene function

```


<br />
<br />

##### Volcano Plot
```{r}
#reset par
par(mfrow=c(1,1))
# Make a basic volcano plot
with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-3,3)))

# Add colored points: blue if padj<0.01, red if log2FC>1 and padj<0.05)
with(subset(res, padj<.01 ), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, padj<.01 & abs(log2FoldChange)>2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
```

<br />
<br />

##### PCA
```{r}
#First we need to transform the raw count data
#vst function will perform variance stabilizing transformation

vsdata <- vst(dds, blind=FALSE)

```


```{r}
plotPCA(vsdata, intgroup="dex") #using the DESEQ2 plotPCA fxn we can
                                #look at how our samples group by treatment
```

